neo/Ledger/MemPool.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Neo.Network.P2P.Payloads;
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+
+namespace Neo.Ledger
+{
+    internal class MemPool : IReadOnlyCollection<Transaction>
+    {
+        private class PoolItem
+        {
+            public readonly Transaction Transaction;
+            public readonly DateTime Timestamp;
+
+            public PoolItem(Transaction tx)
+            {
+                Transaction = tx;
+                Timestamp = DateTime.UtcNow;
+            }
+        }
+
+        private const int MemoryPoolSize = 50000;
+        private readonly ConcurrentDictionary<UInt256, PoolItem> mem_pool = new ConcurrentDictionary<UInt256, PoolItem>();
+
+        public int Count => mem_pool.Count;
+
+        public void Clear()
+        {
+            mem_pool.Clear();
+        }
+
+        public bool ContainsKey(UInt256 hash)
+        {
+            return mem_pool.ContainsKey(hash);
+        }
+
+        public IEnumerator<Transaction> GetEnumerator()
+        {
+            return mem_pool.Select(p => p.Value.Transaction).GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public void RemoveLowestFee(int count)
+        {
+            if (count <= 0) return;
+            if (count >= mem_pool.Count)
+            {
+                mem_pool.Clear();
+            }
+            else
+            {
+                UInt256[] delete = mem_pool.AsParallel()
+                    .OrderBy(p => p.Value.Transaction.NetworkFee / p.Value.Transaction.Size)
+                    .ThenBy(p => p.Value.Transaction.NetworkFee)
+                    .ThenBy(p => new BigInteger(p.Key.ToArray()))
+                    .Take(count)
+                    .Select(p => p.Key)
+                    .ToArray();
+                foreach (UInt256 hash in delete)
+                    mem_pool.TryRemove(hash, out _);
+            }
+        }
+
+        public void RemoveOldFree(DateTime time)
+        {
+            UInt256[] hashes = mem_pool
+                .Where(p => p.Value.Timestamp < time && p.Value.Transaction.NetworkFee == Fixed8.Zero)
+                .Select(p => p.Key)
+                .ToArray();
+            foreach (UInt256 hash in hashes)
+                mem_pool.TryRemove(hash, out _);
+        }
+
+        public bool TryAdd(UInt256 hash, Transaction tx)
+        {
+            mem_pool.TryAdd(hash, new PoolItem(tx));
+
+            if (mem_pool.Count > MemoryPoolSize)
+            {
+                RemoveOldFree(DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 20));

```

Comment:
Once the mempool doesn't have any transactions that are free, but the pool is full this will repeatedly be called, maybe we could keep track of the lowest fee in the pool to know whether this should be called every time at transaction is added in the case that the pool is full of transactions that all have > 0 fee.

Reply:
We can store a flag of how many fee tx are inside for call this method or not

Reply:
Or we can create two pools, one high priority and the other low priority.

Reply:
I like the solution with a separate data structure for the free transactions.